### PR TITLE
Allow waiters to introspect last result

### PR DESCRIPTION
**Pull request by harlowja**
This will help implement the desire listed at issue #101
by allowing for waiter functions/classes/other to recieve
the last result and do smart things with it.

### CODE REVIEW

**Comment by harlowja** [comment]
This doesn't need to be the way we do this, but this is one of such ways...

File: tenacity/__init__.py [hunk 1]
```diff
@@ -251,8 +251,15 @@ def iter(self, result, exc_info, start_time):
...
         if self.wait:
-            sleep = self.wait(self.statistics['attempt_number'],
-                              delay_since_first_attempt)
+            # This will allow for passing in the result and handling
+            # the older versions of these functions that do not take
+            # the prior result.
+            try:
+                sleep = self.wait(self.statistics['attempt_number'],
+                                  delay_since_first_attempt, last_result=fut)
+            except TypeError:
```
Comment (harlowja) [inline tenacity/__init__.py]: I don't like this capture here.... other option would be using `isinstance` of a new baseclass...

Comment (jd) [reply tenacity/__init__.py]: did you try to `inspect` the signature?

Comment (harlowja) [reply tenacity/__init__.py]: I didn't though that should work.

File: tenacity/__init__.py [hunk 2]
```diff
@@ -80,6 +82,15 @@
 from .after import after_nothing  # noqa
 
 
+def _waiter_takes_last_result(waiter):
```
Comment (jd) [inline tenacity/__init__.py]: this could/should be a method

File: tenacity/__init__.py [hunk 3]
```diff
@@ -80,6 +82,15 @@
...
+    if not six.callable(waiter):
+        return False
+    if isinstance(waiter, _wait.wait_base):
+        waiter = waiter.__call__
```
Comment (jd) [inline tenacity/__init__.py]: Not sure that's gonna work for code that inherits from `wait_base` but is not updated.

Comment (harlowja) [reply tenacity/__init__.py]: Seems like it should work?

```
>>> from tenacity import wait
>>> class OldWait(wait.wait_base):
...    def __call__(self, attempt, delay):
...        pass
...
>>> my_wait = OldWait()
>>>
>>> import inspect
>>> inspect.getargspec(my_wait.__call__)
ArgSpec(args=['self', 'attempt', 'delay'], varargs=None, keywords=None, defaults=None)
```

File: tenacity/__init__.py (hunk 2 above)
Comment (harlowja) [reply tenacity/__init__.py]: Done

File: tenacity/__init__.py (hunk 3 above)
Comment (jd) [reply tenacity/__init__.py]: Sorry I misread the line 89. I read `waiter = wait_base.__call__`. Facepalm.

**Pull request merged**